File: common/src/main/java/com/pluxity/permission/dto/PermissionGroupResponse.java
package com.pluxity.permission.dto;

import com.fasterxml.jackson.annotation.JsonUnwrapped;
import com.pluxity.global.response.BaseResponse;
import com.pluxity.permission.Permission;
import com.pluxity.permission.PermissionGroup;
import io.swagger.v3.oas.annotations.media.Schema;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public record PermissionGroupResponse(
        @Schema(description = "권한 그룹 ID") Long id,
        @Schema(description = "권한 그룹 이름") String name,
        @Schema(description = "권한 그룹 설명") String description,
        @Schema(description = "포함된 권한 목록 ") List<PermissionResponse> permissions,
        @JsonUnwrapped BaseResponse baseResponse) {

    public static PermissionGroupResponse from(PermissionGroup permissionGroup) {
        Map<String, List<Permission>> groupedPermissions =
                permissionGroup.getPermissions().stream()
                        .collect(Collectors.groupingBy(Permission::getResourceName));

        List<PermissionResponse> permissionDtos =
                groupedPermissions.entrySet().stream()
                        .map(
                                entry -> {
                                    String resourceType = entry.getKey();
                                    List<String> resourceIds =
                                            entry.getValue().stream()
                                                    .map(Permission::getResourceId)
                                                    .collect(Collectors.toList());
                                    return new PermissionResponse(resourceType, resourceIds);
                                })
                        .toList();

        return new PermissionGroupResponse(
                permissionGroup.getId(),
                permissionGroup.getName(),
                permissionGroup.getDescription(),
                permissionDtos,
                BaseResponse.of(permissionGroup));
    }
}

File: common/src/main/java/com/pluxity/permission/dto/PermissionGroupResponse.kt
package com.pluxity.permission.dto

import com.fasterxml.jackson.annotation.JsonUnwrapped
import com.pluxity.global.response.BaseResponse
import com.pluxity.permission.PermissionGroup
import io.swagger.v3.oas.annotations.media.Schema
import kotlin.collections.component1
import kotlin.collections.component2

data class PermissionGroupResponse(
    @field:Schema(description = "권한 그룹 ID")
    val id: Long,
    @field:Schema(description = "권한 그룹 이름")
    val name: String,
    @field:Schema(description = "권한 그룹 설명")
    val description: String?,
    @field:Schema(description = "포함된 권한 목록 ")
    val permissions: List<PermissionResponse>,
    @field:JsonUnwrapped
    val baseResponse: BaseResponse?,
)

fun PermissionGroup.toPermissionGroupResponse(): PermissionGroupResponse =
    PermissionGroupResponse(
        id = this.id!!,
        name = this.name,
        description = this.description,
        permissions =
            this.permissions.groupBy { it.resourceName }.map { (resourceType, permissions) ->
                PermissionResponse(
                    resourceType = resourceType,
                    resourceIds = permissions.map { it.resourceId },
                )
            },
        baseResponse = BaseResponse.of(this),
    )
